File: src/backend/app/schemas/dispensas.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.solicitud_dispensa import EstadoSolicitud
from app.schemas.grados import GradoOut


class AlumnoMinOut(BaseModel):
    """Datos mínimos del alumno propietario embebidos en la solicitud."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    username: str
    nombre: str
    apellidos: str


class ResponsableMinOut(BaseModel):
    """Datos mínimos del usuario que tomó la solicitud para revisión."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    nombre: str
    apellidos: str


class AsignaturaEmbedOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    codigo: str
    nombre: str
    ects: float
    caracter: str
    curso_plan: int
    grados: list[GradoOut]


class AsignaturaMatriculadaEmbedOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    n_matricula: int
    asignatura: AsignaturaEmbedOut


class SolicitudDispensaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    alumno: AlumnoMinOut
    asignatura_matriculada: AsignaturaMatriculadaEmbedOut
    motivo: str | None
    estado: EstadoSolicitud
    observaciones: str | None
    fecha_solicitud: datetime
    fecha_resolucion: datetime | None
    responsable: ResponsableMinOut | None


class CrearSolicitudRequest(BaseModel):
    """Alta de una solicitud.

    Para el Alumno: `alumno_id` se descarta (se resuelve desde la sesión, patrón
    propietario implícito).

    Para la Secretaria: `alumno_id` debe venir explícito (opera en nombre de).
    """

    model_config = ConfigDict(extra="ignore")

    alumno_id: int | None = None
    asignatura_matriculada_id: int
    motivo: str


class EditarSolicitudRequest(BaseModel):
    """PATCH unificado para Alumno, Secretaria y Director.

    Todos los campos son opcionales (`None` = no tocar). La `PoliticaAcceso`
    aplicable según el rol decide qué se permite tocar:

      - Alumno (sólo PENDIENTE):     motivo, asignatura_matriculada_id + estado→ANULADA
      - Secretaria (sólo PENDIENTE): motivo, asignatura_matriculada_id + estado→ANULADA
      - Director:                    observaciones + estado→EN_REVISION|APROBADA|RECHAZADA
    """

    model_config = ConfigDict(extra="ignore")

    estado: EstadoSolicitud | None = None
    motivo: str | None = None
    asignatura_matriculada_id: int | None = None
    observaciones: str | None = None


class ExportarDispensasFiltros(BaseModel):
    """Filtros opcionales del export CSV.

    Como query params en el endpoint — `BaseModel` sólo documenta el contrato.
    """

    estado: EstadoSolicitud | None = None
    alumno_id: int | None = None
    desde: datetime | None = None
    hasta: datetime | None = None

File: src/frontend/src/types/dispensas.ts
import type { Grado } from './grados';

export type EstadoSolicitud =
  | 'pendiente'
  | 'en_revision'
  | 'aprobada'
  | 'rechazada'
  | 'anulada';

export const ESTADOS_TERMINALES: ReadonlySet<EstadoSolicitud> = new Set([
  'aprobada',
  'rechazada',
  'anulada',
]);

export interface AlumnoMin {
  id: number;
  username: string;
  nombre: string;
  apellidos: string;
}

export interface ResponsableMin {
  id: number;
  nombre: string;
  apellidos: string;
}

export interface AsignaturaEmbed {
  id: number;
  codigo: string;
  nombre: string;
  ects: number;
  caracter: 'OB' | 'OP' | 'FB';
  curso_plan: number;
  grados: Grado[];
}

export interface AsignaturaMatriculadaEmbed {
  id: number;
  n_matricula: number;
  asignatura: AsignaturaEmbed;
}

export interface SolicitudDispensa {
  id: number;
  alumno: AlumnoMin;
  asignatura_matriculada: AsignaturaMatriculadaEmbed;
  motivo: string | null;
  estado: EstadoSolicitud;
  observaciones: string | null;
  fecha_solicitud: string;
  fecha_resolucion: string | null;
  responsable: ResponsableMin | null;
}

export interface CrearSolicitudRequest {
  alumno_id?: number;
  asignatura_matriculada_id: number;
  motivo: string;
}

export interface EditarSolicitudRequest {
  estado?: EstadoSolicitud;
  motivo?: string;
  asignatura_matriculada_id?: number;
  observaciones?: string;
}

export interface FiltrosDispensa {
  estado?: EstadoSolicitud;
  alumno_id?: number;
  desde?: string;
  hasta?: string;
}
